File: BackEnd/PHP/laravelshop/app/Modules/Ranking/Services/RankingService.php
<?php

namespace App\Modules\Ranking\Services;

use App\Common\Exceptions\BusinessException;
use App\Models\Product;
use App\Models\RecentProductView;
use App\Models\SearchHistory;
use App\Models\User;
use Illuminate\Support\Facades\DB;
use Symfony\Component\HttpFoundation\Response;

class RankingService
{
    public function popularProducts(?int $categoryId = null, int $limit = 20): array
    {
        $limit = min(100, max(1, $limit));
        $query = RecentProductView::query()
            ->select('product_id', DB::raw('COUNT(*) as score'))
            ->groupBy('product_id')
            ->orderByDesc('score')
            ->limit($limit);

        if ($categoryId) {
            $query->whereHas('product', fn ($builder) => $builder->where('category_id', $categoryId));
        }

        return $query->get()->values()->map(function ($row, int $index): array {
            $product = Product::query()->with('priceEntries')->find($row->product_id);

            return [
                'rank' => $index + 1,
                'rankChange' => 0,
                'product' => $product ? [
                    'id' => $product->id,
                    'name' => $product->name,
                    'lowestPrice' => (float) ($product->priceEntries->min('price') ?? 0),
                    'thumbnailUrl' => $product->thumbnail_url,
                ] : null,
                'score' => (int) $row->score,
            ];
        })->all();
    }

    public function popularKeywords(int $limit = 20): array
    {
        $limit = min(100, max(1, $limit));

        return SearchHistory::query()
            ->select('keyword', DB::raw('COUNT(*) as search_count'))
            ->groupBy('keyword')
            ->orderByDesc('search_count')
            ->limit($limit)
            ->get()
            ->values()
            ->map(fn ($row, int $index): array => [
                'rank' => $index + 1,
                'keyword' => $row->keyword,
                'searchCount' => (int) $row->search_count,
                'rankChange' => 0,
            ])->all();
    }

    public function recalculate(User $actor): array
    {
        if ($actor->role !== 'ADMIN') {
            throw new BusinessException('관리자 권한이 필요합니다.', 'FORBIDDEN', Response::HTTP_FORBIDDEN);
        }

        $updatedCount = RecentProductView::query()->distinct('product_id')->count('product_id');

        return [
            'updatedCount' => $updatedCount,
        ];
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/ranking/RankingService.java
package com.pbshop.java.spring.maven.jpa.postgresql.ranking;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pbshop.java.spring.maven.jpa.postgresql.activity.RecentProductView;
import com.pbshop.java.spring.maven.jpa.postgresql.activity.RecentProductViewRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.activity.SearchHistory;
import com.pbshop.java.spring.maven.jpa.postgresql.activity.SearchHistoryRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.BusinessException;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.ErrorCode;
import com.pbshop.java.spring.maven.jpa.postgresql.product.PriceEntry;

@Service
@Transactional(readOnly = true)
public class RankingService {

    private final RecentProductViewRepository recentProductViewRepository;
    private final SearchHistoryRepository searchHistoryRepository;

    public RankingService(
            RecentProductViewRepository recentProductViewRepository,
            SearchHistoryRepository searchHistoryRepository
    ) {
        this.recentProductViewRepository = recentProductViewRepository;
        this.searchHistoryRepository = searchHistoryRepository;
    }

    public List<Map<String, Object>> getPopularProducts(Long categoryId, Integer limit) {
        int size = normalizeLimit(limit);
        AtomicInteger rank = new AtomicInteger(1);

        return recentProductViewRepository.findAll().stream()
                .filter(view -> categoryId == null || view.getProduct().getCategory().getId().equals(categoryId))
                .collect(Collectors.groupingBy(RecentProductView::getProduct, Collectors.counting()))
                .entrySet()
                .stream()
                .sorted((left, right) -> Long.compare(right.getValue(), left.getValue()))
                .limit(size)
                .map(entry -> {
                    List<PriceEntry> prices = entry.getKey().getPriceEntries();
                    BigDecimal lowestPrice = prices.stream()
                            .map(PriceEntry::getPrice)
                            .min(BigDecimal::compareTo)
                            .orElse(BigDecimal.ZERO);
                    Map<String, Object> response = new LinkedHashMap<>();
                    response.put("rank", rank.getAndIncrement());
                    response.put("rankChange", 0);
                    response.put("product", Map.of(
                            "id", entry.getKey().getId(),
                            "name", entry.getKey().getName(),
                            "lowestPrice", lowestPrice.setScale(2, RoundingMode.HALF_UP),
                            "thumbnailUrl", entry.getKey().getThumbnailUrl() == null ? "" : entry.getKey().getThumbnailUrl()
                    ));
                    response.put("score", entry.getValue());
                    return response;
                })
                .toList();
    }

    public List<Map<String, Object>> getPopularKeywords(Integer limit) {
        int size = normalizeLimit(limit);
        AtomicInteger rank = new AtomicInteger(1);

        return searchHistoryRepository.findAll().stream()
                .collect(Collectors.groupingBy(SearchHistory::getKeyword, Collectors.counting()))
                .entrySet()
                .stream()
                .sorted((left, right) -> Long.compare(right.getValue(), left.getValue()))
                .limit(size)
                .map(entry -> {
                    Map<String, Object> response = new LinkedHashMap<>();
                    response.put("rank", rank.getAndIncrement());
                    response.put("keyword", entry.getKey());
                    response.put("searchCount", entry.getValue());
                    response.put("rankChange", 0);
                    return response;
                })
                .toList();
    }

    public Map<String, Object> recalculate(AuthenticatedUserPrincipal principal) {
        requireAdmin(principal);
        long productCount = recentProductViewRepository.findAll().stream()
                .map(RecentProductView::getProduct)
                .map(product -> product.getId())
                .distinct()
                .count();
        long keywordCount = searchHistoryRepository.findAll().stream()
                .map(SearchHistory::getKeyword)
                .distinct()
                .count();
        return Map.of("updatedCount", productCount + keywordCount);
    }

    private int normalizeLimit(Integer limit) {
        if (limit == null || limit <= 0) {
            return 20;
        }
        return Math.min(limit, 100);
    }

    private void requireAdmin(AuthenticatedUserPrincipal principal) {
        if (principal == null) {
            throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다.");
        }
        if (!"ADMIN".equalsIgnoreCase(principal.role())) {
            throw new BusinessException(ErrorCode.FORBIDDEN, "관리자 권한이 필요합니다.");
        }
    }
}
